### PR TITLE
Update sampling API according to the spec

### PR DESCRIPTION
**Pull request by lmolkova**
This change
- allows passing attributes during span creation
- changes `ISampler` to abstract `Sampler` class and allows passing attributes to `ShouldSample`
- cleans up static factories in favor of constructors on samplers implementations

Is seems there are more changes coming to the spec (but not in the spec yet) from [otep](https://github.com/open-telemetry/oteps/blob/master/text/0006-sampling.md) to support sampling hint - this PR does **not** implement them.

There is no clarity in the spec how sampler should use attributes and whether it needs to return them back in the decision. There is no info in the spec on how Span should use attributes from samplers. Other implementations (checked python and java) are not consistent. So I'll leave it to the spec to clarify it.

For now, let's assume attributes could only be used to make the decision and this change simply enable this scenario.

This is a first step to fix https://github.com/open-telemetry/opentelemetry-dotnet/issues/273

### CODE REVIEW

File: src/OpenTelemetry/Trace/Span.cs [hunk 1]
```diff
@@ -106,6 +107,15 @@ public sealed class Span : ISpan, IDisposable
             if (this.IsRecording)
             {
                 this.SetLinks(links);
+
+                if (spanCreationOptions?.Attributes != null)
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry/Trace/Span.cs]: yay!

File: src/OpenTelemetry/Trace/Samplers/ProbabilitySampler.cs [hunk 2]
```diff
@@ -18,55 +18,47 @@
...
             else
             {
-                idUpperBound = (long)(probability * long.MaxValue);
+                this.IdUpperBound = (long)(probability * long.MaxValue);
             }
-
-            return new ProbabilitySampler(probability, idUpperBound);
         }
 
+        public override string Description { get; }
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Samplers/ProbabilitySampler.cs]: Nice catch!

After thinking more about it - I don't think `Probability` or moreover, `IdUpperBound` need to be public at all. 

Users configured `probability` and they should know it. There is no obvious scenario now why they would need to get this from the sampler.
`IdUpperBound` is aux field and it's derived from the probability and max long.

I'll keep them private for now and if we'll ever find where they need to be exposed - happy to return them to public surface.

[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply src/OpenTelemetry/Trace/Samplers/ProbabilitySampler.cs]: Description is often used to "export" the current configuration. Like what @austinlparker asking for here: https://github.com/open-telemetry/opentelemetry-specification/issues/286

[the remark this answers is not in the document]
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Samplers/ProbabilitySampler.cs]: Right, Description stays. It's `Probability` and `IdUpperBound` I made private. Did I miss something?

**Pull request merged**